cpp/include/raft/sparse/matrix/preprocessing.cuh

```diff
@@ -0,0 +1,562 @@
+ * @param[in] full_id_len
+ *   Value that represents the number overall number of features observed during the fit
+ * cycle.
+ * @param[in] num_feats
+ *   Value that represents the number of features that exist for the matrices encoded.
+ * */
+template <typename ValueType, typename IndexType>
+SparseEncoder<ValueType, IndexType>::SparseEncoder(
+  raft::resources& handle, int* featIdValues, int num_feats, int num_rows, int full_id_len)
+  : numFeats(num_feats), numRows(num_rows), fullIdLen(full_id_len)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  cudaMallocManaged(&featIdCount, numFeats * sizeof(int));
+  cudaMemset(featIdCount, 0, numFeats * sizeof(int));
+  for (int i = 0; i < numFeats; i++) {
+    featIdCount[i] = featIdValues[i];
+  }
+}
+
+/**
+ * This destructor deallocates/frees the reserved memory of the class.
+ *
+ * @tparam ValueType
+ *   Type of the values in the sparse matrix.
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ * */
+template <typename ValueType, typename IndexType>
+SparseEncoder<ValueType, IndexType>::~SparseEncoder()
+{
+  cudaFree(featIdCount);
+}
+
+/**
+ * This function exports all values required to recreate the SparseEncoder to a
+ * file.
+ *
+ * @tparam ValueType
+ *   Type of the values in the sparse matrix.
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ *
+ * @param[in] handle
+ *   Container for managing reusable resources.
+ * @param[in] save_path
+ *   The path to save the file container all values required to recreate SparseEncoder.
+ * */
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::save(raft::resources& handle, std::string save_path)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  auto featIdCount_md = raft::make_device_vector<IndexType, int64_t>(handle, numFeats);
+  raft::copy(featIdCount_md.data_handle(), featIdCount, numFeats, stream);
+  std::ofstream saveFile(save_path);
+  if (saveFile.is_open()) {
+    std::ostringstream oss;
+    saveFile << numFeats << " ";
+    saveFile << numRows << " ";
+    saveFile << fullIdLen << " ";
+    for (int i = 0; i < numFeats; i++) {
+      saveFile << featIdCount[i] << " ";
+    }
+    saveFile.close();
+  }
+}
+
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::_fit_feats(IndexType* cols,
+                                                     IndexType* counts,
+                                                     IndexType nnz,
+                                                     IndexType* results)
+{
+  int blockSize  = (nnz < 256) ? nnz : 256;
+  int num_blocks = (nnz + blockSize - 1) / blockSize;
+  raft::sparse::matrix::detail::_scan<<<blockSize, num_blocks>>>(cols, nnz, counts);
+  raft::sparse::matrix::detail::_fit_compute_occurs<<<blockSize, num_blocks>>>(
+    cols, nnz, counts, results, numFeats);
```

Comment:
I think we can probably simplify this - 

This seems like this is doing something like `np.bincount`  - counting up the document frequencies of each term so that we can compute the `idf` for tf-idf/bm25.

I think we can avoid having the _scan/_fit_compute_occurs kernels here entirely (and also remove the requirement for the `coo_sort` that is performed previously) by using cub.  For example - other code inside of raft is using cub and `HistogramEven` to peform the bincount operation  (like https://github.com/rapidsai/raft/blob/842afd7e5f6309a1094d2790f2ec50594306d490/cpp/include/raft/stats/detail/silhouette_score.cuh#L98-L146 etc)